In [21]:
import csv
f = open("guns.csv") # Create File handler csv

# Use the csv.reader() and list() methods to read the file into a list
guns = list(csv.reader(f))

# Select  1st 5 rows in guns, except the header
headers=guns[0]
guns_data = guns[1:]
headers, guns_data[0:5]

(['',
  'year',
  'month',
  'intent',
  'police',
  'sex',
  'age',
  'race',
  'hispanic',
  'place',
  'education'],
 [['1',
   '2012',
   '01',
   'Suicide',
   '0',
   'M',
   '34',
   'Asian/Pacific Islander',
   '100',
   'Home',
   '4'],
  ['2',
   '2012',
   '01',
   'Suicide',
   '0',
   'F',
   '21',
   'White',
   '100',
   'Street',
   '3'],
  ['3',
   '2012',
   '01',
   'Suicide',
   '0',
   'M',
   '60',
   'White',
   '100',
   'Other specified',
   '4'],
  ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'],
  ['5',
   '2012',
   '02',
   'Suicide',
   '0',
   'M',
   '31',
   'White',
   '100',
   'Other specified',
   '2']])

In [22]:
year_counts = {}

for row in guns_data:
    row_year = row[1]
    if row_year in year_counts:
        year_counts[row_year] = year_counts[row_year]+1
    else:
        year_counts[row_year] = 1   
year_counts

{'2012': 33563, '2013': 33636, '2014': 33599}

In [23]:
import datetime
date_counts = {}

for row in guns_data:
    dates = datetime.datetime(year=int(row[1]), month=int(row[2]), day=1)
    if dates in date_counts:
        date_counts[dates] = date_counts[dates]+1
    else:
        date_counts[dates] = 1
        
date_counts

{datetime.datetime(2012, 1, 1, 0, 0): 2758,
 datetime.datetime(2012, 2, 1, 0, 0): 2357,
 datetime.datetime(2012, 3, 1, 0, 0): 2743,
 datetime.datetime(2012, 4, 1, 0, 0): 2795,
 datetime.datetime(2012, 5, 1, 0, 0): 2999,
 datetime.datetime(2012, 6, 1, 0, 0): 2826,
 datetime.datetime(2012, 7, 1, 0, 0): 3026,
 datetime.datetime(2012, 8, 1, 0, 0): 2954,
 datetime.datetime(2012, 9, 1, 0, 0): 2852,
 datetime.datetime(2012, 10, 1, 0, 0): 2733,
 datetime.datetime(2012, 11, 1, 0, 0): 2729,
 datetime.datetime(2012, 12, 1, 0, 0): 2791,
 datetime.datetime(2013, 1, 1, 0, 0): 2864,
 datetime.datetime(2013, 2, 1, 0, 0): 2375,
 datetime.datetime(2013, 3, 1, 0, 0): 2862,
 datetime.datetime(2013, 4, 1, 0, 0): 2798,
 datetime.datetime(2013, 5, 1, 0, 0): 2806,
 datetime.datetime(2013, 6, 1, 0, 0): 2920,
 datetime.datetime(2013, 7, 1, 0, 0): 3079,
 datetime.datetime(2013, 8, 1, 0, 0): 2859,
 datetime.datetime(2013, 9, 1, 0, 0): 2742,
 datetime.datetime(2013, 10, 1, 0, 0): 2808,
 datetime.datetime(2013, 11,

In [24]:
sex_counts = {}

for row in guns_data:
    row_sex = row[5]
    if row_sex in sex_counts:
        sex_counts[row_sex] = sex_counts[row_sex]+1
    else:
        sex_counts[row_sex] = 1   
sex_counts

### OR:
sex_list = [row[5] for row in guns_data]
sex_counts2 = {}
for sex in sex_list:
    if sex in sex_counts2:
        sex_counts2[sex] += 1
    else:
        sex_counts2[sex] = 1
sex_counts,sex_counts2         

({'F': 14449, 'M': 86349}, {'F': 14449, 'M': 86349})

In [25]:
age_counts = {}

for row in guns_data:
    row_age = row[6]
    if row_age in age_counts:
        age_counts[row_age] = age_counts[row_age]+1
    else:
        age_counts[row_age] = 1   
# age_counts

### OR :
age_list = [row[6] for row in guns_data]
age_counts2 = {}
for age in age_list:
    if age in age_counts2:
        age_counts2[age] += 1
    else:
        age_counts2[age] = 1
age_counts,age_counts2       

({'0': 33,
  '1': 38,
  '10': 53,
  '100': 1,
  '101': 2,
  '102': 2,
  '107': 1,
  '11': 61,
  '12': 117,
  '13': 229,
  '14': 364,
  '15': 561,
  '16': 864,
  '17': 1185,
  '18': 1753,
  '19': 2065,
  '2': 50,
  '20': 2219,
  '21': 2504,
  '22': 2712,
  '23': 2472,
  '24': 2437,
  '25': 2230,
  '26': 2231,
  '27': 2070,
  '28': 1986,
  '29': 1955,
  '3': 66,
  '30': 1869,
  '31': 1833,
  '32': 1824,
  '33': 1700,
  '34': 1699,
  '35': 1631,
  '36': 1512,
  '37': 1500,
  '38': 1491,
  '39': 1389,
  '4': 54,
  '40': 1414,
  '41': 1485,
  '42': 1492,
  '43': 1527,
  '44': 1449,
  '45': 1372,
  '46': 1437,
  '47': 1532,
  '48': 1621,
  '49': 1669,
  '5': 43,
  '50': 1674,
  '51': 1755,
  '52': 1715,
  '53': 1708,
  '54': 1684,
  '55': 1596,
  '56': 1625,
  '57': 1472,
  '58': 1510,
  '59': 1430,
  '6': 50,
  '60': 1361,
  '61': 1306,
  '62': 1099,
  '63': 1041,
  '64': 1126,
  '65': 1039,
  '66': 998,
  '67': 865,
  '68': 868,
  '69': 879,
  '7': 43,
  '70': 883,
  '71': 791,
  '72': 736

In [26]:
race_counts = {}

for row in guns_data:
    row_race = row[7]
    if row_race in race_counts:
        race_counts[row_race] = race_counts[row_race]+1
    else:
        race_counts[row_race] = 1   
race_counts

{'Asian/Pacific Islander': 1326,
 'Black': 23296,
 'Hispanic': 9022,
 'Native American/Native Alaskan': 917,
 'White': 66237}

highest gun related deaths among whites, aged 19-27 , and males

In [27]:
import csv
c = open("census.csv", "r") # Create File handler csv

# Use the csv.reader() and list() methods to read the file into a list
census = list(csv.reader(c))

# Select  1st 5 rows in guns, except the header
headers2=census[0]
census_data = census[1:]
headers2, census_data

(['Id',
  'Year',
  'Id',
  'Sex',
  'Id',
  'Hispanic Origin',
  'Id',
  'Id2',
  'Geography',
  'Total',
  'Race Alone - White',
  'Race Alone - Hispanic',
  'Race Alone - Black or African American',
  'Race Alone - American Indian and Alaska Native',
  'Race Alone - Asian',
  'Race Alone - Native Hawaiian and Other Pacific Islander',
  'Two or More Races'],
 [['cen42010',
   'April 1, 2010 Census',
   'totsex',
   'Both Sexes',
   'tothisp',
   'Total',
   '0100000US',
   '',
   'United States',
   '308745538',
   '197318956',
   '44618105',
   '40250635',
   '3739506',
   '15159516',
   '674625',
   '6984195']])

In [28]:
mapping = {
    "Black" : census[1][12],
    "Hispanic" : census[1][11],
    "Native American/Native Alaskan" : census[1][13],
    "White" : census[1][10],
    "Asian/Pacific Islander": census[1][14] + census[1][15]+census[1][14] + census[1][16]   
          }
mapping

{'Asian/Pacific Islander': '15159516674625151595166984195',
 'Black': '40250635',
 'Hispanic': '44618105',
 'Native American/Native Alaskan': '3739506',
 'White': '197318956'}

In [29]:
race_per_hundredk ={}
for race, count in race_counts.items():
    race_per_hundredk[race] = (count/int(mapping[race]))*100000
race_per_hundredk    

{'Asian/Pacific Islander': 8.746980714890028e-21,
 'Black': 57.8773477735196,
 'Hispanic': 20.220491210910907,
 'Native American/Native Alaskan': 24.521955573811088,
 'White': 33.56849303419181}

In [30]:
intents = {}

for row in guns_data:
    row_intent = row[3]
    if row_intent in intents:
        intents[row_intent] = intents[row_intent]+1
    else:
        intents[row_intent] = 1   
intents

{'Accidental': 1639,
 'Homicide': 35176,
 'NA': 1,
 'Suicide': 63175,
 'Undetermined': 807}

In [31]:
# re-Extract the intent and race columns using a list comprehension
intents = [row[3] for row in guns_data]
races = [row[7] for row in guns_data]

homicide_race_counts = {}
for i, race in enumerate(races):
    if intents[i] == "Homicide":
        if race in homicide_race_counts:
            homicide_race_counts[race] += 1
        else:
            homicide_race_counts[race] = 1
homicide_race_counts

{'Asian/Pacific Islander': 559,
 'Black': 19510,
 'Hispanic': 5634,
 'Native American/Native Alaskan': 326,
 'White': 9147}

Homicides numbers is highest among blacks, followed by and doubling the number among whites. Numbers alone are not sufficiient to analyze. We need the proportion of each race in the US to fully compare the numbers above. Thus we want want to get a rate of gun deaths/100000 people of each race. In order to do this, we need to read in data about what percentage of the US population falls into each racial category.

In [32]:
homicide_race_per_100k = {}
for race, count in homicide_race_counts.items():
    homicide_race_per_100k[race] = count/int(mapping[race])*100000
homicide_race_per_100k

{'Asian/Pacific Islander': 3.6874526543163845e-21,
 'Black': 48.471284987180944,
 'Hispanic': 12.627161104219914,
 'Native American/Native Alaskan': 8.717729026240365,
 'White': 4.6356417981453335}

Thus we see death by homicides in Blacks is 12x higher than in whites, and 4x higher than hispanics

In [33]:
# Analyze link, if any, between month and homicide rate.
months = [row[2] for row in guns_data]
homicide_month_counts = {}

for i, mon in enumerate(months):
    if intents[i] == "Homicide":
        if mon in homicide_month_counts:
            homicide_month_counts[mon] += 1
        else:
            homicide_month_counts[mon] = 1

for mon, cnt in homicide_month_counts.items():
    print(mon, cnt, "High" if cnt>=3000 else "Low")


01 2829 Low
02 2178 Low
04 2845 Low
11 2919 Low
06 3130 High
09 2966 Low
03 2780 Low
07 3269 High
12 3191 High
08 3125 High
10 2968 Low
05 2976 Low


Highest homicide rates occur in Summer June-August, and in December. Lowest rates are in February. 

In [34]:
# Analyze link, if any, homicide rate by gender.
##sex_count = [row[5] for row in guns_data] ##(alreadt computed above)
homicide_sex_counts = {}

for i, sex in enumerate(sex_counts):
    if intents[i] == "Homicide":
        if sex in homicide_sex_counts:
            homicide_sex_counts[sex] += 1
        else:
            homicide_sex_counts[sex] = 1

homicide_sex_counts

{}

men commit homicides almost 6x more than women

In [35]:
suicide_sex_counts = {}

for i, sex in enumerate(sex_counts):
    if intents[i] == "Suicide":
        if sex in suicide_sex_counts:
            suicide_sex_counts[sex] += 1
        else:
            suicide_sex_counts[sex] = 1

suicide_sex_counts

{'F': 1, 'M': 1}

In [36]:
Men commit Suicides almost 6x more than women

SyntaxError: invalid syntax (<ipython-input-36-bec072c627c3>, line 1)

In [ ]:
suicide_race_counts = {}
for i, race in enumerate(races):
    if intents[i] == "Suicide":
        if race in suicide_race_counts:
            suicide_race_counts[race] += 1
        else:
            suicide_race_counts[race] = 1
##print(suicide_race_counts)

suicide_race_per_hundredk ={}
for race, count in suicide_race_counts.items():
    suicide_race_per_hundredk[race] = (count/int(mapping[race]))*100000
suicide_race_counts, suicide_race_per_hundredk

Suicide is the highest among Whites. Lowest among Asians/pacific

In [ ]:
accident_race_counts = {}
for i, race in enumerate(races):
    if intents[i] == "Accidental":
        if race in accident_race_counts:
            accident_race_counts[race] += 1
        else:
            accident_race_counts[race] = 1
##print(accident_race_counts)

accident_race_per_hundredk ={}
for race, count in accident_race_counts.items():
    accident_race_per_hundredk[race] = (count/int(mapping[race]))*100000
accident_race_counts, accident_race_per_hundredk

Accidental Gun Shots is the highest among Black people.

In [ ]:
location_list = [row[9] for row in guns_data]
location_counts = {}
for i, location in enumerate(location_list):
    #if intents[i] == "Accidental":
        if location+" "+intents[i] in location_counts:
            location_counts[location+" "+intents[i] ] += 1
        else:
            location_counts[location+" "+intents[i] ] = 1
location_counts

The majority of gun-related deaths happen at home by by suicides 
followed by homicides.  Next highest are Street Homicides. 

 'Street Homicide'	8,867
 'Home Homicide'	13,613
 'Home Suicide'	45,415


In [37]:
education = {
    1 : "Less than High School",
    2 : "Graduated from High School or equivalent",
    3 : "Some College",
    4 : "At least graduated from College",
    5 : "Not available"   
          }

education_list = [row[10] for row in guns_data]

education_counts = {}
for i, education in enumerate(education_list):
    if education+" "+intents[i] in education_counts:
        education_counts[education+" "+intents[i] ] += 1
    else:
        education_counts[education+" "+intents[i] ] = 1
        
education_counts

{'1 Accidental': 492,
 '1 Homicide': 11839,
 '1 Suicide': 9292,
 '1 Undetermined': 200,
 '2 Accidental': 633,
 '2 Homicide': 15649,
 '2 Suicide': 26321,
 '2 Undetermined': 324,
 '3 Accidental': 327,
 '3 Homicide': 5640,
 '3 Suicide': 15533,
 '3 Undetermined': 180,
 '4 Accidental': 146,
 '4 Homicide': 1559,
 '4 NA': 1,
 '4 Suicide': 11147,
 '4 Undetermined': 93,
 '5 Accidental': 28,
 '5 Homicide': 455,
 '5 Suicide': 876,
 '5 Undetermined': 10,
 'NA Accidental': 13,
 'NA Homicide': 34,
 'NA Suicide': 6}

  '3 Suicide'	15533
  '2 Homicide'	15649
  '2 Suicide'	26321
The highest mumbers of gun related deaths are among those who "Graduated from High School or equivalent" by suicides or bt Homicide. - THis is followed closedly by suicides among those who have some college education.

In [62]:
education_def = {
    1 : "Less than High School",
    2 : "Graduated from High School or equivalent",
    3 : "Some College",
    4 : "At least graduated from College",
    5 : "Not available"   
          }

education_list = [row[10] for row in guns_data]

education_counts = {}
for i, education in enumerate(education_list):
    if education =="NA":
        ed_index=5
    else:
        ed_index=int(education)
    education_val=(education_def[ed_index])
    if education_val+" "+intents[i] in education_counts:
        education_counts[education_val+" "+intents[i] ] += 1
    else:
        education_counts[education_val+" "+intents[i] ] = 1
education_counts

{'At least graduated from College Accidental': 146,
 'At least graduated from College Homicide': 1559,
 'At least graduated from College NA': 1,
 'At least graduated from College Suicide': 11147,
 'At least graduated from College Undetermined': 93,
 'Graduated from High School or equivalent Accidental': 633,
 'Graduated from High School or equivalent Homicide': 15649,
 'Graduated from High School or equivalent Suicide': 26321,
 'Graduated from High School or equivalent Undetermined': 324,
 'Less than High School Accidental': 492,
 'Less than High School Homicide': 11839,
 'Less than High School Suicide': 9292,
 'Less than High School Undetermined': 200,
 'Not available Accidental': 41,
 'Not available Homicide': 489,
 'Not available Suicide': 882,
 'Not available Undetermined': 10,
 'Some College Accidental': 327,
 'Some College Homicide': 5640,
 'Some College Suicide': 15533,
 'Some College Undetermined': 180}

In [72]:
import operator
sorted_education_counts = sorted(education_counts.items(), key=operator.itemgetter(1), reverse=True)    
sorted_education_counts

[('Graduated from High School or equivalent Suicide', 26321),
 ('Graduated from High School or equivalent Homicide', 15649),
 ('Some College Suicide', 15533),
 ('Less than High School Homicide', 11839),
 ('At least graduated from College Suicide', 11147),
 ('Less than High School Suicide', 9292),
 ('Some College Homicide', 5640),
 ('At least graduated from College Homicide', 1559),
 ('Not available Suicide', 882),
 ('Graduated from High School or equivalent Accidental', 633),
 ('Less than High School Accidental', 492),
 ('Not available Homicide', 489),
 ('Some College Accidental', 327),
 ('Graduated from High School or equivalent Undetermined', 324),
 ('Less than High School Undetermined', 200),
 ('Some College Undetermined', 180),
 ('At least graduated from College Accidental', 146),
 ('At least graduated from College Undetermined', 93),
 ('Not available Accidental', 41),
 ('Not available Undetermined', 10),
 ('At least graduated from College NA', 1)]

 'Graduated from High School or equivalent Suicide', 26321)
 
 'Graduated from High School or equivalent Homicide', 15649)
 
 'Some College Suicide', 15533),